### ISSUE TITLE
Torch performance issues

### ISSUE THREAD

**Issue by lukeyeager** [NVIDIA/DIGITS#339]
Breaking discussion from https://github.com/NVIDIA/DIGITS/pull/324#issuecomment-142373731 out into a separate issue.

> @lukeyeager - I just wanted to make a note that a training job which takes 6sec with Caffe takes 38sec with Torch.

> @gheinrich - The difference between Torch and Caffe should be less dramatic on "bigger" models.

>You're right. Here's the results of a [very much non-rigorous] test I tried. It's one epoch on a dataset of ~40k images.

> . | AlexNet (sec) | GoogLeNet (sec)
------------ | ------------- | -------------
Caffe (v0.13.2, with cuDNN v3 and CNMeM) | 126 | 722
Torch (cuDNN v3) | 252 | 969
*Slowdown* | 2x | 1.3x

**Comment by gheinrich** [comment]
cc @soumith as he expressed interest in this.

I have pushed a couple of changes [there](https://github.com/gheinrich/DIGITS/commits/dev/torch-speed) which speed Torch training up (figures below are for 20 epochs of training LeNet on MNIST 45k training samples and 15k validation samples, with one validation per epoch):
fa00a945198c00a189601d3bb8713b009b05f11a removes some unnecessary computations of train set accuracy on CPU side: 187s -> 142s
7635864105af716e45bf79382fac8db7093512b4 implements a multi-threaded data loader for LMDB: 142s -> 83s
The same benchmark takes 73s with Caffe so there is still some room for improvement.
For example I don't understand why training is slower when I copy tensors to GPU from the data loader threads (as opposed to doing it from the main thread).
The current version greedily allocates new tensors for each training example. It might speed things up to re-use a small number of tensors across batches.

**Comment by soumith** [comment]
nice.

Could you point me to instructions on how to create the imagenet lmdb? I'm trying to get things up and running, but that's kind of a blocker for me at the moment...

Should I use caffe's imagenet instructions? https://github.com/BVLC/caffe/tree/master/examples/imagenet

**Comment by borisfom** [comment]
Thanks Soumith!

I have submitted PR#64 with decently complete cud_v4 bindings - please take a look:
https://github.com/soumith/cudnn.torch/pull/63

I will merge it later today if there are no objections.

Best,
-Boris.

**Comment by borisfom** [comment]
Correction - had to resubmit PR against the right branch (R4), here:
https://github.com/soumith/cudnn.torch/pull/64


From: Boris Fomitchev <bfomitchev@nvidia.com<mailto:bfomitchev@nvidia.com>>
Date: Friday, October 9, 2015 at 11:39 AM
To: NVIDIA/DIGITS <reply@reply.github.com<mailto:reply@reply.github.com>>, NVIDIA/DIGITS <DIGITS@noreply.github.com<mailto:DIGITS@noreply.github.com>>
Subject: Re: [DIGITS] Torch performance issues (#339)

Thanks Soumith!

I have submitted PR#64 with decently complete cud_v4 bindings - please take a look:
https://github.com/soumith/cudnn.torch/pull/63

I will merge it later today if there are no objections.

Best,
-Boris.

**Comment by gheinrich** [comment]
@soumith 

> Could you point me to instructions on how to create the imagenet lmdb?

You need to create the LMDB using DIGITS if you want to use DIGITS to train a model. This [page](
https://github.com/NVIDIA/DIGITS/blob/master/docs/ImageFolderFormat.md) explains how to structure your image folders in a way that DIGITS can understand. There is a section on this page that explains how to create a subset of imagenet (it takes a while to create the full imagenet LMDB).

This [page](https://github.com/NVIDIA/DIGITS/blob/master/docs/GettingStarted.md#creating-a-dataset) then explains how to create the LMDB using DIGITS. Let us know if you need any more information. Thanks!

**Comment by borisfom** [comment]
This is now merged into master cudnn.torch/R4, thanks Soumith!

Best,
-Boris.

**Comment by gheinrich** [comment]
I don't think there is a general performance issue with the integration of Torch into DIGITS anymore. Some models train faster with Torch, other models train faster with Caffe. Some numbers below:

- LeNet on MNIST (30 epochs):

Number of GPUs  | Caffe | Torch
------------- | ------------- | ----------
1  | 55s | 56s
2  | 1m6s | 1m27s
(training is slower with multiple GPUs presumably due to the communication overhead)

- Alexnet on upscaled `256x256` CIFAR10 (5 epochs):

Number of GPUs  | Caffe | Torch
------------- | ------------- | ----------
1  |  9m31s | 6m38s
2  | 7m46s | 5m33s

- GoogleNet on upscaled `256x256` CIFAR10 (1 epoch):

Number of GPUs  | Caffe | Torch
------------- | ------------- | ----------
1  |  4m3s | 11m13s
2  | 3m12s | 7m32s
(Torch slowliness mostly due to extra Batch Normalization layers)

**Comment by soumith** [comment]
Interesting data points. Thanks. Just a note on BatchNorm, latest nn/cunn have super optimized batchnorm (faster than CuDNN R4).
